File: JAVA/src/main/scala/engine/math/Vector2f.scala
package engine.math

/**
  * Created by Mnenmenth Alkaborin
  * Please refer to LICENSE file if included
  * for licensing information
  * https://github.com/Mnenmenth
  */

class Vector2f(x1: Float, y1: Float) {

  var x: Float = x1
  var y: Float = y1
  def string(): String = "(x: " + x + ", y: " + y + ")"

  def length(): Float = Math.sqrt(x * x + y * y).toFloat
  def dot(v: Vector2f): Float = x * v.x + y * v.y

  def normalize(): Vector2f = {
    val l = length
    x /= l
    y /= l
    this
  }

  def rotate(angle: Float): Vector2f = {
    val radians: Double = Math.toRadians(angle)
    val cos: Double = Math.cos(radians)
    val sin: Double = Math.sin(radians)
    new Vector2f((x * cos - y * sin).toFloat, (x * sin + y * cos).toFloat)
  }

  def add(v: Vector2f): Vector2f = {
    new Vector2f(x + v.x, y + v.y)
  }

  def add(v: Float): Vector2f = {
    new Vector2f(x + v, y + v)
  }

  def sub(v: Vector2f): Vector2f = {
    new Vector2f(x - v.x, y - v.y)
  }

  def sub(v: Float): Vector2f = {
    new Vector2f(x - v, y - v)
  }

  def mult(v: Vector2f): Vector2f = {
    new Vector2f(x * v.x, y * v.y)
  }

  def mult(v: Float): Vector2f = {
    new Vector2f(x * v, y * v)
  }

  def div(v: Vector2f): Vector2f = {
    new Vector2f(x / v.x, y / v.y)
  }

  def div(v: Float): Vector2f = {
    new Vector2f(x / v, y / v)
  }

}

File: CPP/src/engine/math/Vector2f.cpp
//
// Created by Mnenmenth on 1/7/2016.
//

#include "Vector2f.h"
#include <sstream>
#include <math.h>

Vector2f::Vector2f(float x, float y) {
    Vector2f::x = x;
    Vector2f::y = y;
}

void Vector2f::setX(float x) {
    Vector2f::x = x;
}

void Vector2f::setY(float y) {
    Vector2f::y = y;
}

float Vector2f::getX() {
    return Vector2f::x;
}

float Vector2f::getY() {
    return Vector2f::y;
}

std::string Vector2f::toString() {
    std::stringstream stream;
    stream << "(x: " << Vector2f::x << ", y: " << Vector2f::y << ")";
    return stream.str();
}

float Vector2f::length() {
    return (float) sqrt(Vector2f::x * Vector2f::x + Vector2f::y * Vector2f::y);
}

float Vector2f::dot(Vector2f v) {
    return Vector2f::x * v.getX() + Vector2f::y * v.getY();
}

Vector2f Vector2f::normalize() {
    float length = Vector2f::length();
    Vector2f::x /= length;
    Vector2f::y /= length;
    return *this;
}

Vector2f Vector2f::rotate(float angle) {
    double radians = angle * (3.141592653 / 180);
    double cos_ = cos(radians);
    double sin_ = sin(radians);
    return Vector2f((float) (Vector2f::x * cos_ - Vector2f::y * sin_),
                    (float) (Vector2f::x * sin_ + Vector2f::y * cos_));
}

Vector2f Vector2f::add(Vector2f v) {
    return Vector2f(Vector2f::x + v.getX(), Vector2f::y + v.getY());
}

Vector2f Vector2f::add(float v) {
    return Vector2f(Vector2f::x + v, Vector2f::y + v);
}

Vector2f Vector2f::sub(Vector2f v) {
    return Vector2f(Vector2f::x - v.getX(), Vector2f::y - v.getY());
}

Vector2f Vector2f::sub(float v) {
    return Vector2f(Vector2f::x - v, Vector2f::y - v);
}

Vector2f Vector2f::mult(Vector2f v) {
    return Vector2f(Vector2f::x * v.getX(), Vector2f::y * v.getY());
}

Vector2f Vector2f::mult(float v) {
    return Vector2f(Vector2f::x * v, Vector2f::y * v);
}

Vector2f Vector2f::div(Vector2f v) {
    return Vector2f(Vector2f::x / v.getX(), Vector2f::y / v.getY());
}

Vector2f Vector2f::div(float v) {
    return Vector2f(Vector2f::x / v, Vector2f::y / v);
}